### ISSUE TITLE
Disappearing Inventory

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1819]
_Imported from trac ticket 1819.
Created by **mariam**
Opened in trac: Jul 23, 2012, 7:38:33 AM
Last modified in trac: Feb 11, 2013, 8:00:01 AM_
***
_by mariam:_
Kent,

You may recall an issue we were having with inventory that ‘disappeared’ from the sales orders.

You put a control in place in Loft to manage changes that occurred on both GERS and Loft at the same time.

We continue to have this issue.

Order:                   0714210LE18

SKU:                      642706608

Orig. Location:   10 TERRAC

 

ITQ:

Show OOR transaction reserving the inventory from 10 TERRAC  but no UNR transaction and the location is now null.

There’s nothing in the comments that hints any manipulation of sku 642706608.

 

If we would have encountered this problem before we rolled out Loft, I would not be so persistent, but we didn’t.

I will create an OVR transaction to allow the store to reserve the line and finalize the order because they can’t wait.

Any thoughts?

Mariam H. Farach
----
If you _never_ noticed this before Loft, then it certainly seems to be related to Loft.  That doesn't tell us which system made the error, only that it seems to be either Loft or the effect of Loft modified data upon GERS.  As far as I can tell, Loft will never unreserve a line without also writing the transaction and the inventory to the database.

GERS is undoubtedly programmed the same way, but with the difference that Loft is extremely strict at writing /atomic transactions /while past experience indicates to me that GERS is very sloppy with transactional control.  In other words, it is feasible, for example, that GERS was used to unreserve the line; it did so, committed the SO_LN changes to the database (naughty) and then, while going to write the inventory and transaction, encountered some error (maybe caused by a Loft side effect of data) and that portion of the event failed.

It's also possible that Loft created the error.  We are only keeping a couple GB of logging, so the log files don't go back more than a couple days. I just changed that /times 10/ this morning, so if this happens again, we have a better chance of having log files going back far enough.

Here are the recent inventory transactions:  It doesn't tell us much, but it's worth noting that the original reservation was made from GERS (source was SOM).

```
TYP PRODUCTID        QTY OL OLDLOC NE NEWLOC ORDERID        OVE SOURC CREATEDATE           USERNAME
--- --------- ---------- -- ------ -- ------ -------------- --- ----- -------------------- ----------
IUR 642706608          1           10 STORE                     ISTM  24-JAN-2011 22:47:17 LISUSER
ITF 642706608          1 00 WHSE   10 STORE                     ISTM  24-JAN-2011 22:47:17 LISUSER
COL 642706608          1 10 STORE                               COL   25-JAN-2011 13:45:52 HEJ
COL 642706608          1           10 TERRAC                    COL   25-JAN-2011 13:45:52 HEJ
OOR 642706608          1 10 TERRAC           0714210LE18        SOM   14-JUL-2012 18:29:55 YAV4317
OVR 642706608          1           10 TERRAC                ADJ OVR   20-JUL-2012 18:01:50 MAHC
OOR 642706608          1 10 TERRAC           0714210LE18        SOM   20-JUL-2012 18:02:11 MAHC
```
Couple tests:

I'm very curious what would happen if you unreserved that line right now in GERS to see what would happen.  Does it work properly?
If so, reserve it again from SOM and unreserve it from Loft (leave the order open in SOM while doing this in Loft).

Another test:

1. Unfill the line.
2. In screen A, open the order in SOM (with that line unfilled).
3. In screen B, reserve the line.
4. Now, back to screen A, make some other change to that line (maybe the price temporarily or percent of sale or sometime).
5. What happens?

**Comment by kbower** [comment]
_**kbower** trac comment on Jul 23, 2012, 7:41:52 AM:_

May need to wait for more data...

**Comment by kbower** [comment]
_**kbower** trac comment on Aug 1, 2012, 10:09:39 AM:_

Background: every SQL statement Loft issues is logged. Every single one.  We have all logs going back to at least July 21, including the creation of these 4 order lines on July 30 at 5:24:11. 

I've searched every occurrence of every order line update statement and there are none referencing this order (0730210PP85). 

What this confirms is that Loft did not update those order lines; I'm certain. (Recall my earlier claim that Loft always writes an unfill transaction and writes inventory in the same database transaction as the unfill of the line itself, so we already claimed it wasn't Loft, but this confirms that.)

Possibilities:
1. Gers wrote the OOR transaction but did not update the so_ln. (it may have hit an error). Do we know that the lines showed as reserved after the SOM fill by YAV4317 on July 30 at 10:20 or are we just assuming that?
2. GERS unfilled the line but failed to write inventory or the transaction. 
3.  The nightly unfill report or some other report unfilled the line without writing the transaction or inventory. 
4. Others?

Do not take this personally, for it isn't at all, but of these possibilities, #3 seems most suspicious to me.  If (and only if) that nightly report is one written by El Dorado, I wouldn't mind looking at it. Is there a commit between the So_ln update and the writing of inventory and transaction?  If there is one, it must be removed. If there is an exception block that is catching oracle errors, is it masking problems? (It is almost always worse to ignore problems than allow them to raise exception).

Let us know,
Kent
